MudDataGrid: Fix Filter Panel Positioning with FixedHeader = true

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Filter Panel when DataGridFilterMode.Simple is placed in <tbody> which scrolls normally even under Sticky Header conditions. I was not able to relocate the components to any other part of the table without ill side effects and none that rendered the panel where it needed to be. Added js exception handling in mudPopover.js to relocate any filters-panel to the first headercell in a table when opened.
Resolves #10282 

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Visual Tests, new component to test with as well as going through entire doc page for DataGrid testing.
![filterfix](https://github.com/user-attachments/assets/7dcdd4e1-ca7d-4580-9d6c-7d51b009df59)

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

versile2 committed:
Added exception handling for filters-panel due to it can't be put in the appropriate location in razor markup

ScarletKuro commented:
https://github.com/MudBlazor/MudBlazor/blob/fdf2f459d8d3b78c6bb43d3fd716c4ea46d49b4e/src/MudBlazor/TScripts/mudPopover.js#L346
The only way to fix this would be to add an exception case in `mudPopover.js`. If there are no other good alternatives, I probably wouldn’t fix it at all. I don’t think it's ideal to apply patches to individual components in popover. I made an exception for MudList (https://github.com/MudBlazor/MudBlazor/pull/10216) because the popover menu with list is a logical part of the popover, but the filter panel is not.

@henon @danielchalmers, thoughts?

versile2 commented:
> https://github.com/MudBlazor/MudBlazor/blob/fdf2f459d8d3b78c6bb43d3fd716c4ea46d49b4e/src/MudBlazor/TScripts/mudPopover.js#L346
> 
> 
> The only way to fix this would be to add an exception case in `mudPopover.js`? If there are no other good alternatives, I probably wouldn’t fix it at all. I don’t think it's ideal to apply patches to individual components in popover. I made an exception for MudList (#10216) because the popover menu with list is a logical part of the popover, but the filter panel is not.
> @henon @danielchalmers, thoughts?

I may have found a razor solution. Working on it now.

versile2 committed:
switched to a non js solution

ScarletKuro commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor:
```diff
@@ -58,7 +58,36 @@
-                        
+                                @if (_filtersMenuVisible && FilterMode == DataGridFilterMode.Simple)
+                                {
+                                <MudOverlay ZIndex="12" @bind-Visible="@_filtersMenuVisible" AutoClose OnClosed="@CloseFilters"></MudOverlay>
```
Incorrect indent after `{`.

ScarletKuro commented:
> I may have found a razor solution. Working on it now.

Now it looks good to me

versile2 commented:
> > I may have found a razor solution. Working on it now.
> 
> Now it looks good to me

It does push down the first row when you open the filter menu now. It's not bad but it is there. I can't seem to remove it. And in it's new position it does it for all.

versile2 committed:
indent fix

danielchalmers reviewed (changes requested):
Shouldn't shift rows

versile2 commented:
> Shouldn't shift rows

I will look again, not sure I can keep it from shifting rows. The existing rows are solid, but also dynamic so I couldn't add it there. It needs a parent to relate to so I added the row. When scrolling it takes the row to the "top" of the grid which moves the popover to the top of the grid. I added a top: 58px to push it down (which is pushing the row). I'll work on it some more this weekend.

versile2 committed:
Merge branch 'MudBlazor:dev' into fix/datagridfixedheaderfilter

versile2 committed:
fix for row spacing

versile2 reviewed (commented):
No longer shifts rows

![Screenshot 2024-11-24 194020](https://github.com/user-attachments/assets/3588ea89-9caf-4c30-9dd5-7056ab6b92c2)
![Screenshot 2024-11-24 194119](https://github.com/user-attachments/assets/e485e571-2af6-4519-a12a-22ce7225c06b)

![image](https://github.com/user-attachments/assets/56a3560d-ec57-4d77-922a-6cfb2bf1fa92)

versile2 commented:
> Shouldn't shift rows

it was as simple as setting visibility: collapse instead of hidden. I'll take it. Checked all datagrids (with filter simple) in datagrid examples as well as the test subject.

henon commented:
Thanks @versile2

Dzubrul commented:
Thank you @versile2 !

RJ-Ponder commented:
Hi @versile2, thanks for this work! I was looking out for a solution to this bug for a while.

When I tried it out, the behavior is a little different than expected. When scrolled down, I expect the filter pop-up to show below the header like it does when scrolled up and like your screenshots show. Instead, it covers the header. Is this something that can be fixed?

versile2 commented:
I don't believe this PR is in Version 7, in V8 it works perfectly. You can try it out and see if your issue is fixed at 

https://try.arctechonline.tech 

Note the try I host does not save snippets so you will have to copy/paste code.

RJ-Ponder commented:
I should have mentioned I am using V8. The code snippet I'm testing with is from the test component in your PR:
`src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridFixedHeaderFilterTest.razor`

Interesting, I see it does work correctly in the playground, but the exact snippet in my project works differently. I'm on .NET 8 though. I'll see if I can find if it's another library or something interfering with the css.

versile2 commented:
> I should have mentioned I am using V8. The code snippet I'm testing with is from the test component in your PR: `src/MudBlazor.UnitTests.Viewer/TestComponents/DataGrid/DataGridFixedHeaderFilterTest.razor`
> 
> Interesting, I see it does work correctly in the playground, but the exact snippet in my project works differently. I'm on .NET 8 though. I'll see if I can find if it's another library or something interfering with the css.

Make sure to refresh css files and/or try in "private or incognito" mode to force reload of css. I've had that issue myself with this version and so many style changes.

versile2 commented:
Also that's preview 6 or 7 I'm hosting (I don't remember which block I copied that from but the datetime stamp is correct)

RJ-Ponder commented:
Oops, turns out it was some custom css I added a while back and forgot about. After getting rid of that it works perfectly. Thanks again for the fix!

versile2 commented:
> Oops, turns out it was some custom css I added a while back and forgot about. After getting rid of that it works perfectly. Thanks again for the fix!

I suspect there might be some of that with V8, so many of the popover issues will be fixed. Thanks for the feedback, hopefully it's all good but don't hesitate if it's not.

RJ-Ponder commented:
One small thing I noticed is that it doesn't take into account Dense="true". The shorter row height causes a gap between the header and filter popover.

versile2 commented:
> @versile2 I found that the same problem exists with the Column Panel menu. I can open a new issue, but wanted to comment here first because I think the fix will be similar to the filter panel.
> 
> See example here (only difference from the example in this PR is the ShowMenuIcon option) https://try.mudblazor.com/snippet/mYGJklEVLygUcXHl
> 
> The behavior is the same in V8, the link is just so I could save the snippet easily.

Make a new issue if you could. Tag me in it. Also include #10292 so I can reference back here and the Dense problem if it still exists. 
https://try.arctechonline.tech should have the latest dev changes as of yesterday if you want to test anything.
